keras_nlp/samplers/contrastive_sampler.py

```diff
@@ -0,0 +1,146 @@
+    def __init__(
+        self,
+        k=5,
+        alpha=0.5,
+        seed=None,
+    ):
+        super().__init__()
+        self.k = k
+        self.alpha = alpha
+        self.seed = seed
+
+    def __call__(
+        self,
+        next,
+        prompt,
+        state=None,
+        index=0,
+        mask=None,
+        end_token_id=None,
+    ):
+        batch_size, max_length = tf.shape(prompt)[0], tf.shape(prompt)[1]
+        # Make sure max length and start index are the same dtype.
+        index = tf.cast(index, max_length.dtype)
+
+        def create_beams(x):
+            """Add initial beam state."""
+            x = tf.repeat(x, self.k, axis=0)
+            flat_shape = [batch_size * self.k] + x.shape.as_list()[1:]
+            return tf.reshape(x, shape=flat_shape)
+
+        def flatten_beams(x):
+            """Combine the beam dim and batch dim."""
+            flat_shape = [batch_size * self.k] + x.shape.as_list()[2:]
+            return tf.reshape(x, shape=flat_shape)
+
+        def unflatten_beams(x):
+            """Separate the beam dim and batch dim."""
+            unflat_shape = [batch_size, self.k] + x.shape.as_list()[1:]
+            return tf.reshape(x, shape=unflat_shape)
+
+        mask = tf.zeros_like(prompt, dtype=tf.bool) if mask is None else mask
+        # `tf.while_loop` will not accept `None` as a value for `loop_vars`.
+        state = () if state is None else state
+
+        def cond(prompt, state, index):
+            if end_token_id is None:
+                return True
+            # Stop if all sequences have produced a *new* end_token_id.
+            end_tokens = (prompt == end_token_id) & (~mask)
+            prompt_done = tf.reduce_any(end_tokens, axis=-1)
+            return not tf.reduce_all(prompt_done)
+
+        def body(prompt, state, index):
+            # Compute the softmax distribution for the next token.
+            logits, state = next(prompt, state, index)
+            probabilities = keras.activations.softmax(logits)
+
+            prompt_beams, mask_beams = create_beams(prompt), create_beams(mask)
+            state_beams = tf.nest.map_structure(create_beams, state)
+
+            top_k_probabilities, top_k_indices = tf.math.top_k(
+                probabilities, k=self.k, sorted=False
+            )
+            next_token_probabilities = flatten_beams(top_k_probabilities)
+            next_token = flatten_beams(top_k_indices)
+            next_token = tf.cast(next_token, prompt.dtype)
+
+            next_token = tf.where(
+                mask_beams[:, index], prompt_beams[:, index], next_token
+            )
+            # Update the prompt with the next token.
+            next_token = next_token[:, tf.newaxis]
+            prompt_beams = dynamic_update_slice(
+                prompt_beams, next_token, [0, index]
+            )
+
+            _, next_state_beams = next(prompt_beams, state_beams, index + 1)
+            hidden_states = next_state_beams["hidden_states"]
```

Comment:
ideally we can avoid indexing into `state` here.

I think we probably want to have a `next` callback always return `hidden_states` and probably will need to pass the initial hidden_state tensor to the sampler in the first place.